### PR TITLE
Synchronize Http/2 HPack implementation between CoreFx and ASP.NET Core

### PR DESCRIPTION
**Pull request by jkotalik**
Part of https://github.com/aspnet/AspNetCore/issues/13900. Corresponding PRs: https://github.com/dotnet/corefx/pull/41069, https://github.com/dotnet/corefx/pull/42070

This PR synchronizes the HTTP/2 HPack implementation between Kestrel and CoreFx. Eventually, we will find a mechanism for sharing this code, but this is a necessary step in creating a shared source. The only remaining differences between aspnetcore and corefx are exceptions and namespaces.

This PR changes the following functionality:
- Changes the IntegerDecoding logic
- Adds a bunch of static methods to HPackEncoder (unused for now).

### CODE REVIEW

**Comment by jkotalik** [comment]
Going to update this PR now as https://github.com/dotnet/corefx/pull/41069 is merged.

**Comment by jkotalik** [comment]
@Tratcher is this ready for review?

**Comment by Tratcher** [comment]
No, I still need to get the tests passing. Feel free to look over the general structure though.

**Comment by Tratcher** [comment]
Tests are passing. The main work left now is to deduplicate any test coverage and see if Kestrel has any more to add to the shared tests.

**Comment by Tratcher** [comment]
This is now ready for review.

Please remember that any comments on the shared code should be made in the corefx PR. https://github.com/dotnet/corefx/pull/42070

File: src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.netcoreapp.cs [hunk 1]
```diff
@@ -204,14 +204,6 @@ public enum HttpMethod : byte
         Custom = (byte)9,
         None = (byte)255,
     }
-    public partial class HttpParser<TRequestHandler> : Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.IHttpParser<TRequestHandler> where TRequestHandler : Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.IHttpHeadersHandler, Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.IHttpRequestLineHandler
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.netcoreapp.cs]: Removing these I think breaks benchmark apps @sebastienros to confirm

Comment (jkotalik) [reply src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.netcoreapp.cs]: I think you may need to update the benchmark apps with new namespaces.

Comment (jkotalik) [reply src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.netcoreapp.cs]: Actually I'm not sure at all. I don't see any of these calls in the benchmark repo.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs [hunk 2]
```diff
@@ -108,7 +108,10 @@ public static unsafe string GetHeaderName(this Span<byte> span)
             return asciiString;
         }
 
-        public static unsafe string GetAsciiStringNonNullCharacters(this Span<byte> span)
+        public static string GetAsciiStringNonNullCharacters(this Span<byte> span)
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs]: Why do we need both the span and readonly span versions of these?

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs]: Was trying to minimize code churn. The HTTP1 parser uses Span in many places, where HTTP2 uses ReadOnlySpan most of the time.

File: src/Shared/Http2/CopyToCoreFx.cmd [hunk 3]
```diff
@@ -0,0 +1,12 @@
+@ECHO OFF
+SETLOCAL
+
+IF "%COREFX_REPO%" == "" (
```
Comment (jkotalik) [inline src/Shared/Http2/CopyToCoreFx.cmd]: These scripts should be able to take an absolute path to the place to copy to, not just an environment variable.

File: src/Shared/Http2/Hpack/HPackDecodingException.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0.
```
Comment (jkotalik) [inline src/Shared/Http2/Hpack/HPackDecodingException.cs]: Add an exclusion to the license.txt file for this folder using a different license.

File: src/Shared/Http2/CopyToCoreFx.cmd (hunk 3 above)
Comment (Tratcher) [reply src/Shared/Http2/CopyToCoreFx.cmd]: What's the usage scenario for an absolute path?

The env variable is for inner loop development.

Comment (jkotalik) [reply src/Shared/Http2/CopyToCoreFx.cmd]: I have multiple dotnet/aspnetcore repos I copy to and from. I have at least two clones of aspnetcore I work on regularly. Also it's more analogous to cp commands in unix.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs (hunk 2 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs]: kk that's fine.

File: src/Shared/Http2/Hpack/HPackDecodingException.cs (hunk 4 above)
Comment (Tratcher) [reply src/Shared/Http2/Hpack/HPackDecodingException.cs]: Do we have an example for that? I don't see any other exclusions.

Comment (jkotalik) [reply src/Shared/Http2/Hpack/HPackDecodingException.cs]: Actually I think you are good, David added the exclusion in his last commit: https://github.com/aspnet/AspNetCore/commit/bd65275148abc9b07a3b59797a88d485341152bf

**Review by halter73 — APPROVED** [review]

**Pull request merged**